File: daemon/src/session-pool.ts
import { ChildProcess, spawn } from "child_process";
import { createInterface } from "readline";
import { v4 as uuidv4 } from "uuid";
import { existsSync } from "fs";
import {
  ManagedSession,
  SessionStatus,
  PermissionMode,
  StreamEvent,
  SessionInfo,
  ClaudeStdoutMessage,
  modeToCliFlag,
} from "./types";
import { MessageQueue } from "./message-queue";

/**
 * Session state — no longer holds a long-lived process.
 * Instead, each send() spawns a new `claude --print` process
 * and uses --resume to maintain conversation context.
 */
interface InternalSession extends ManagedSession {
  /** Currently running Claude process (only during message processing) */
  process: ChildProcess | null;
  queue: MessageQueue;
  /** Whether we're currently processing a message */
  processing: boolean;
  /** Model name reported by Claude CLI (extracted from init message) */
  model: string | null;
}

/**
 * SessionPool manages Claude CLI sessions using per-message spawn.
 *
 * Architecture change (2026-03-14):
 * - Claude CLI 2.1.76 does not support --input-format stream-json without --print
 * - We now spawn a fresh `claude --print <msg> --output-format stream-json` per message
 * - Session continuity is maintained via --resume <sdkSessionId>
 * - Each process lives only for the duration of one message exchange
 */
export class SessionPool {
  private sessions: Map<string, InternalSession> = new Map();

  /**
   * Create a new session (lightweight — just registers session state).
   * No Claude CLI process is spawned until a message is sent.
   */
  async create(path: string, mode: PermissionMode = "auto"): Promise<string> {
    // Validate path exists
    if (!existsSync(path)) {
      throw new Error(`Path does not exist: ${path}`);
    }

    const sessionId = uuidv4();

    console.log(`[SessionPool] Creating session ${sessionId} at ${path} (mode=${mode})`);

    const session: InternalSession = {
      sessionId,
      path,
      mode,
      status: "idle",
      sdkSessionId: null,
      createdAt: new Date(),
      lastActivityAt: new Date(),
      process: null,
      queue: new MessageQueue(),
      processing: false,
      model: null,
    };

    this.sessions.set(sessionId, session);
    return sessionId;
  }

  /**
   * Send a message to a session.
   * Spawns a new Claude CLI process for this message.
   * Returns an async iterable of stream events.
   */
  async *send(
    sessionId: string,
    message: string
  ): AsyncGenerator<StreamEvent, void, unknown> {
    const session = this.getSession(sessionId);

    // If Claude is busy, queue the message
    if (session.processing) {
      const position = session.queue.enqueueUser(message);
      yield { type: "queued", position };
      return;
    }

    // Process this message
    yield* this.processMessage(session, message);
  }

  /**
   * Internal: process a single message by spawning a Claude CLI process.
   *
   * Invocation: claude --print <message> --output-format stream-json --verbose
   *             [--resume <sdkSessionId>] [--dangerously-skip-permissions]
   */
  private async *processMessage(
    session: InternalSession,
    message: string
  ): AsyncGenerator<StreamEvent, void, unknown> {
    session.processing = true;
    session.status = "busy";
    session.lastActivityAt = new Date();

    // Build CLI arguments
    const args = [
      "--print",
      message,
      "--output-format",
      "stream-json",
      "--verbose",
      ...modeToCliFlag(session.mode),
    ];

    // Resume from previous conversation if we have a session ID
    if (session.sdkSessionId) {
      args.push("--resume", session.sdkSessionId);
    }

    console.log(`[SessionPool] Spawning claude for session ${session.sessionId}`);
    console.log(`[SessionPool] Command: claude ${args.map(a => a.includes(' ') ? `"${a}"` : a).join(" ")}`);
    console.log(`[SessionPool] CWD: ${session.path}`);

    // Spawn Claude CLI process for this single message
    const child = spawn("claude", args, {
      cwd: session.path,
      stdio: ["pipe", "pipe", "pipe"],
      env: {
        ...process.env,
        TERM: "dumb",
      },
    });

    session.process = child;

    // Event queue for streaming events from stdout
    const eventQueue: StreamEvent[] = [];
    let resolveWait: (() => void) | null = null;
    let done = false;

    const pushEvent = (event: StreamEvent) => {
      eventQueue.push(event);
      if (resolveWait) {
        resolveWait();
        resolveWait = null;
      }
    };

    // Set up stdout line reader (JSON-lines from Claude CLI)
    if (child.stdout) {
      const stdoutReader = createInterface({
        input: child.stdout,
        crlfDelay: Infinity,
      });

      stdoutReader.on("line", (line: string) => {
        let parsed: ClaudeStdoutMessage;
        try {
          parsed = JSON.parse(line);
        } catch {
          console.log(`[Session ${session.sessionId}] non-JSON stdout: ${line}`);
          return;
        }

        // Extract model name from system init message
        if (parsed.type === "system" && parsed.subtype === "init") {
          const raw = parsed as any;
          if (raw.model) {
            session.model = raw.model;
            console.log(`[Session ${session.sessionId}] Model: ${raw.model}`);
          }
        }

        const event = this.convertToStreamEvent(parsed);
        // Capture SDK session ID
        if (event.session_id) {
          session.sdkSessionId = event.session_id;
        }

        pushEvent(event);
      });
    }

    // Handle stderr
    if (child.stderr) {
      const stderrReader = createInterface({
        input: child.stderr,
        crlfDelay: Infinity,
      });
      stderrReader.on("line", (line: string) => {
        console.error(`[Session ${session.sessionId}] stderr: ${line}`);
      });
    }

    // Handle process exit
    child.on("exit", (code, signal) => {
      console.log(
        `[Session ${session.sessionId}] Process exited: code=${code}, signal=${signal}`
      );

      // Normal exit (code 0) after --print is expected — it means processing is done.
      // Only emit error for abnormal exits.
      if (code !== 0 && code !== null) {
        pushEvent({
          type: "error",
          message: `Claude process exited abnormally (code=${code}, signal=${signal})`,
        });
      }

      done = true;
      // Wake up the yield loop if it's waiting
      if (resolveWait) {
        resolveWait();
        resolveWait = null;
      }
    });

    child.on("error", (err) => {
      console.error(`[Session ${session.sessionId}] Process error:`, err);
      pushEvent({
        type: "error",
        message: `Claude process error: ${err.message}`,
      });
      done = true;
      if (resolveWait) {
        resolveWait();
        resolveWait = null;
      }
    });

    // Close stdin immediately — --print mode reads the prompt from args, not stdin
    if (child.stdin) {
      child.stdin.end();
    }

    try {
      // Yield events as they arrive
      while (true) {
        if (eventQueue.length > 0) {
          const event = eventQueue.shift()!;
          yield event;

          // Terminal events
          if (event.type === "result" || event.type === "error" || event.type === "interrupted") {
            break;
          }
        } else if (done) {
          // Process exited and queue is empty
          break;
        } else {
          // Wait for next event or process exit
          await new Promise<void>((resolve) => {
            resolveWait = resolve;
          });
        }
      }
    } finally {
      // Cleanup
      session.process = null;
      session.processing = false;
      session.status = "idle";

      // Kill process if still alive (e.g. on error/interrupt)
      if (child && !child.killed) {
        try {
          child.kill("SIGTERM");
          setTimeout(() => {
            if (!child.killed) child.kill("SIGKILL");
          }, 3000);
        } catch {
          // ignore
        }
      }

      // Process next queued message if any
      if (session.queue.hasUserPending() && session.status === "idle") {
        const next = session.queue.dequeueUser();
        if (next) {
          this.processQueuedMessage(session, next.message);
        }
      }
    }
  }

  /**
   * Process a queued message in the background.
   */
  private async processQueuedMessage(
    session: InternalSession,
    message: string
  ): Promise<void> {
    try {
      const gen = this.processMessage(session, message);
      for await (const event of gen) {
        if (!session.queue.clientConnected) {
          session.queue.bufferResponse(event);
        }
        // Otherwise events go to stream listeners (if any are attached via server SSE)
      }
    } catch (err) {
      console.error(
        `[Session ${session.sessionId}] Error processing queued message:`,
        err
      );
    }
  }

  /**
   * Convert Claude CLI stdout JSON to our StreamEvent format
   */
  private convertToStreamEvent(msg: ClaudeStdoutMessage): StreamEvent {
    switch (msg.type) {
      case "system":
        return {
          type: "system",
          subtype: msg.subtype,
          session_id: msg.session_id,
          model: (msg as any).model,
          raw: msg,
        };

      case "assistant":
        // Extract text content from assistant message
        if (msg.message?.content) {
          const textBlocks = msg.message.content.filter(
            (b) => b.type === "text"
          );
          const toolBlocks = msg.message.content.filter(
            (b) => b.type === "tool_use"
          );

          if (toolBlocks.length > 0) {
            return {
              type: "tool_use",
              tool: toolBlocks[0].name,
              input: toolBlocks[0].input,
              raw: msg,
            };
          }

          if (textBlocks.length > 0) {
            return {
              type: "text",
              content: textBlocks.map((b) => b.text).join(""),
              raw: msg,
            };
          }
        }
        return { type: "text", content: "", raw: msg };

      case "stream_event":
        // Handle streaming deltas
        if (msg.event?.type === "content_block_delta") {
          if (msg.event.delta?.text) {
            return {
              type: "partial",
              content: msg.event.delta.text,
            };
          }
          if (msg.event.delta?.partial_json) {
            return {
              type: "partial",
              content: msg.event.delta.partial_json,
            };
          }
        }
        if (msg.event?.type === "content_block_start") {
          if (msg.event.content_block?.type === "tool_use") {
            return {
              type: "tool_use",
              tool: msg.event.content_block.name,
              raw: msg,
            };
          }
        }
        return { type: "partial", content: "", raw: msg };

      case "tool_progress":
        return {
          type: "tool_use",
          tool: msg.tool_name,
          message: msg.status,
          raw: msg,
        };

      case "result":
        return {
          type: "result",
          session_id: msg.session_id,
          raw: msg,
        };

      default:
        return { type: "system", raw: msg };
    }
  }

  /**
   * Resume a session — in per-message mode, just update the sdkSessionId.
   * The next send() will use --resume with this ID.
   */
  async resume(
    sessionId: string,
    sdkSessionId?: string
  ): Promise<{ ok: boolean; fallback: boolean; newSessionId?: string }> {
    const session = this.sessions.get(sessionId);

    if (session) {
      if (sdkSessionId) {
        session.sdkSessionId = sdkSessionId;
      }
      session.queue.onClientReconnect();
      return { ok: true, fallback: false };
    }

    return { ok: false, fallback: false };
  }

  /**
   * Destroy a session: kill any running Claude process and clean up
   */
  async destroy(sessionId: string): Promise<boolean> {
    const session = this.sessions.get(sessionId);
    if (!session) return false;

    // Kill any running process
    if (session.process && !session.process.killed) {
      session.process.kill("SIGTERM");
      setTimeout(() => {
        if (session.process && !session.process.killed) {
          session.process.kill("SIGKILL");
        }
      }, 5000);
    }

    session.status = "destroyed";
    session.queue.clear();

    this.sessions.delete(sessionId);
    console.log(`[SessionPool] Destroyed session ${sessionId}`);
    return true;
  }

  /**
   * Set the permission mode for a session.
   * In per-message mode, just update the mode — next spawn will use it.
   */
  async setMode(
    sessionId: string,
    mode: PermissionMode
  ): Promise<boolean> {
    const session = this.getSession(sessionId);
    session.mode = mode;
    console.log(`[SessionPool] Mode changed to ${mode} for session ${sessionId}`);
    return true;
  }

  /**
   * Interrupt the current Claude operation for a session.
   * Sends SIGINT to the running Claude CLI process.
   */
  interrupt(sessionId: string): boolean {
    const session = this.sessions.get(sessionId);
    if (!session) {
      throw new Error(`Session not found: ${sessionId}`);
    }

    if (!session.processing || !session.process || session.process.killed) {
      return false;
    }

    console.log(`[SessionPool] Interrupting session ${sessionId}`);
    session.process.kill("SIGTERM"); // Kill the --print process
    session.queue.clear();

    return true;
  }

  /**
   * Get session info
   */
  getSessionInfo(sessionId: string): SessionInfo {
    const session = this.getSession(sessionId);
    return {
      sessionId: session.sessionId,
      path: session.path,
      status: session.status,
      mode: session.mode,
      sdkSessionId: session.sdkSessionId,
      model: session.model,
      createdAt: session.createdAt.toISOString(),
      lastActivityAt: session.lastActivityAt.toISOString(),
    };
  }

  /**
   * List all sessions
   */
  listSessions(): SessionInfo[] {
    return Array.from(this.sessions.values()).map((s) => ({
      sessionId: s.sessionId,
      path: s.path,
      status: s.status,
      mode: s.mode,
      sdkSessionId: s.sdkSessionId,
      model: s.model,
      createdAt: s.createdAt.toISOString(),
      lastActivityAt: s.lastActivityAt.toISOString(),
    }));
  }

  /**
   * Mark client as disconnected for a session (for MQ buffering)
   */
  clientDisconnect(sessionId: string): void {
    const session = this.sessions.get(sessionId);
    if (session) {
      session.queue.onClientDisconnect();
    }
  }

  /**
   * Buffer a single event for a session
   */
  bufferEvent(sessionId: string, event: StreamEvent): void {
    const session = this.sessions.get(sessionId);
    if (session) {
      session.queue.bufferResponse(event, true);
    }
  }

  /**
   * Mark client as reconnected, return buffered events
   */
  clientReconnect(sessionId: string): StreamEvent[] {
    const session = this.sessions.get(sessionId);
    if (session) {
      return session.queue.onClientReconnect();
    }
    return [];
  }

  /**
   * Get queue stats for a session
   */
  getQueueStats(
    sessionId: string
  ): { userPending: number; responsePending: number; clientConnected: boolean } | null {
    const session = this.sessions.get(sessionId);
    if (!session) return null;
    return session.queue.stats();
  }

  /**
   * Destroy all sessions (cleanup on shutdown)
   */
  async destroyAll(): Promise<void> {
    const sessionIds = Array.from(this.sessions.keys());
    await Promise.all(sessionIds.map((id) => this.destroy(id)));
  }

  /**
   * Get a session by ID, throwing if not found
   */
  private getSession(sessionId: string): InternalSession {
    const session = this.sessions.get(sessionId);
    if (!session) {
      throw new Error(`Session not found: ${sessionId}`);
    }
    return session;
  }
}

File: src/session_pool.rs
use std::collections::HashMap;
use std::path::Path;
use std::sync::Arc;

use chrono::Utc;
use serde_json::Value;
use tokio::io::{AsyncBufReadExt, BufReader};
use tokio::process::{Child, Command};
use tokio::sync::{mpsc, Mutex};
use tracing::{error, info};
use uuid::Uuid;

use crate::message_queue::MessageQueue;
use crate::types::{
    convert_claude_message, PermissionMode, QueueStats, SessionInfo, SessionStatus, StreamEvent,
};

/// Internal session state
struct InternalSession {
    session_id: String,
    path: String,
    mode: PermissionMode,
    status: SessionStatus,
    sdk_session_id: Option<String>,
    created_at: chrono::DateTime<Utc>,
    last_activity_at: chrono::DateTime<Utc>,
    /// Currently running Claude process (only during message processing)
    process: Option<Child>,
    queue: MessageQueue,
    /// Whether we're currently processing a message
    processing: bool,
    /// Model name reported by Claude CLI
    model: Option<String>,
}

/// SessionPool manages Claude CLI sessions using per-message spawn.
///
/// Each call to send() spawns a fresh `claude --print <msg> --output-format stream-json`
/// process. Session continuity is maintained via `--resume <sdkSessionId>`.
pub struct SessionPool {
    sessions: Arc<Mutex<HashMap<String, InternalSession>>>,
}

impl SessionPool {
    pub fn new() -> Self {
        Self {
            sessions: Arc::new(Mutex::new(HashMap::new())),
        }
    }

    /// Create a new session (lightweight — just registers session state).
    /// No Claude CLI process is spawned until a message is sent.
    pub async fn create(&self, path: &str, mode: PermissionMode) -> Result<String, String> {
        // Expand ~ to home directory
        let expanded_path = expand_tilde(path);

        // Validate path exists on the remote machine
        if !Path::new(&expanded_path).exists() {
            return Err(format!("Path does not exist on remote machine: {}", expanded_path));
        }

        let session_id = Uuid::new_v4().to_string();
        let now = Utc::now();

        info!(
            "[SessionPool] Creating session {} at {} (mode={:?})",
            session_id, expanded_path, mode
        );

        let session = InternalSession {
            session_id: session_id.clone(),
            path: expanded_path,
            mode,
            status: SessionStatus::Idle,
            sdk_session_id: None,
            created_at: now,
            last_activity_at: now,
            process: None,
            queue: MessageQueue::new(),
            processing: false,
            model: None,
        };

        self.sessions.lock().await.insert(session_id.clone(), session);
        Ok(session_id)
    }

    /// Send a message to a session.
    /// Returns a receiver of stream events.
    pub async fn send(
        &self,
        session_id: &str,
        message: &str,
    ) -> Result<mpsc::Receiver<StreamEvent>, String> {
        let mut sessions = self.sessions.lock().await;
        let session = sessions
            .get_mut(session_id)
            .ok_or_else(|| format!("Session not found: {}", session_id))?;

        // If Claude is busy, queue the message
        if session.processing {
            let position = session.queue.enqueue_user(message.to_string());
            let (tx, rx) = mpsc::channel(1);
            let _ = tx.send(StreamEvent::Queued { position }).await;
            return Ok(rx);
        }

        // Start processing
        let (tx, rx) = mpsc::channel(256);
        let path = session.path.clone();
        let mode = session.mode;
        let sdk_session_id = session.sdk_session_id.clone();

        session.processing = true;
        session.status = SessionStatus::Busy;
        session.last_activity_at = Utc::now();

        let sessions_ref = self.sessions.clone();
        let session_id_owned = session_id.to_string();
        let message_owned = message.to_string();

        // Drop the lock before spawning the task
        drop(sessions);

        // Spawn the processing task
        tokio::spawn(process_message_loop(
            sessions_ref,
            session_id_owned,
            message_owned,
            path,
            mode,
            sdk_session_id,
            tx,
        ));

        Ok(rx)
    }

    /// Resume a session — update the sdkSessionId
    pub async fn resume(
        &self,
        session_id: &str,
        sdk_session_id: Option<String>,
    ) -> serde_json::Value {
        let mut sessions = self.sessions.lock().await;

        if let Some(session) = sessions.get_mut(session_id) {
            if let Some(sid) = sdk_session_id {
                session.sdk_session_id = Some(sid);
            }
            session.queue.on_client_reconnect();
            serde_json::json!({ "ok": true, "fallback": false })
        } else {
            serde_json::json!({ "ok": false, "fallback": false })
        }
    }

    /// Destroy a session: kill any running Claude process and clean up.
    /// Sends SIGTERM first, then SIGKILL after 5 seconds (matching TypeScript behavior).
    pub async fn destroy(&self, session_id: &str) -> bool {
        let mut sessions = self.sessions.lock().await;

        if let Some(mut session) = sessions.remove(session_id) {
            // Kill any running process with SIGTERM + SIGKILL escalation
            if let Some(ref mut child) = session.process {
                send_sigterm_then_sigkill(child, 5000).await;
            }
            session.status = SessionStatus::Destroyed;
            session.queue.clear();
            info!("[SessionPool] Destroyed session {}", session_id);
            true
        } else {
            false
        }
    }

    /// Set the permission mode for a session
    pub async fn set_mode(&self, session_id: &str, mode: PermissionMode) -> Result<bool, String> {
        let mut sessions = self.sessions.lock().await;
        let session = sessions
            .get_mut(session_id)
            .ok_or_else(|| format!("Session not found: {}", session_id))?;
        session.mode = mode;
        info!(
            "[SessionPool] Mode changed to {:?} for session {}",
            mode, session_id
        );
        Ok(true)
    }

    /// Interrupt the current Claude operation for a session.
    /// Sends SIGTERM to the running Claude CLI process (matching TypeScript behavior).
    pub async fn interrupt(&self, session_id: &str) -> Result<bool, String> {
        let mut sessions = self.sessions.lock().await;
        let session = sessions
            .get_mut(session_id)
            .ok_or_else(|| format!("Session not found: {}", session_id))?;

        if !session.processing {
            return Ok(false);
        }

        if let Some(ref child) = session.process {
            info!("[SessionPool] Interrupting session {}", session_id);
            // Send SIGTERM (not SIGKILL) — matches TypeScript behavior
            send_sigterm(child);
            session.queue.clear();
            Ok(true)
        } else {
            Ok(false)
        }
    }

    /// List all sessions
    pub async fn list_sessions(&self) -> Vec<SessionInfo> {
        let sessions = self.sessions.lock().await;
        sessions
            .values()
            .map(|s| SessionInfo {
                session_id: s.session_id.clone(),
                path: s.path.clone(),
                status: s.status,
                mode: s.mode,
                sdk_session_id: s.sdk_session_id.clone(),
                model: s.model.clone(),
                created_at: s.created_at.to_rfc3339(),
                last_activity_at: s.last_activity_at.to_rfc3339(),
            })
            .collect()
    }

    /// Mark client as disconnected for a session
    pub async fn client_disconnect(&self, session_id: &str) {
        let mut sessions = self.sessions.lock().await;
        if let Some(session) = sessions.get_mut(session_id) {
            session.queue.on_client_disconnect();
        }
    }

    /// Buffer a single event for a session
    pub async fn buffer_event(&self, session_id: &str, event: StreamEvent) {
        let mut sessions = self.sessions.lock().await;
        if let Some(session) = sessions.get_mut(session_id) {
            session.queue.buffer_response(event, true);
        }
    }

    /// Mark client as reconnected, return buffered events
    pub async fn client_reconnect(&self, session_id: &str) -> Vec<StreamEvent> {
        let mut sessions = self.sessions.lock().await;
        if let Some(session) = sessions.get_mut(session_id) {
            session.queue.on_client_reconnect()
        } else {
            Vec::new()
        }
    }

    /// Get queue stats for a session
    pub async fn get_queue_stats(&self, session_id: &str) -> Option<QueueStats> {
        let sessions = self.sessions.lock().await;
        sessions.get(session_id).map(|s| s.queue.stats())
    }

    /// Destroy all sessions (cleanup on shutdown)
    pub async fn destroy_all(&self) {
        let mut sessions = self.sessions.lock().await;
        for (id, session) in sessions.iter_mut() {
            if let Some(ref mut child) = session.process {
                send_sigterm_then_sigkill(child, 5000).await;
            }
            info!("[SessionPool] Destroyed session {} (shutdown)", id);
        }
        sessions.clear();
    }
}

// ─── Signal Helpers ───

/// Send SIGTERM to a child process (Unix only).
fn send_sigterm(child: &Child) {
    if let Some(pid) = child.id() {
        #[cfg(unix)]
        {
            use nix::sys::signal::{kill, Signal};
            use nix::unistd::Pid;
            let _ = kill(Pid::from_raw(pid as i32), Signal::SIGTERM);
        }
        #[cfg(not(unix))]
        {
            let _ = pid; // suppress unused warning
        }
    }
}

/// Send SIGTERM, wait `timeout_ms`, then SIGKILL if still alive.
/// Matches TypeScript's `kill("SIGTERM"); setTimeout(() => kill("SIGKILL"), timeout)` pattern.
async fn send_sigterm_then_sigkill(child: &mut Child, timeout_ms: u64) {
    send_sigterm(child);
    tokio::select! {
        _ = child.wait() => {
            // Process exited gracefully after SIGTERM
        }
        _ = tokio::time::sleep(std::time::Duration::from_millis(timeout_ms)) => {
            // Timeout — escalate to SIGKILL
            let _ = child.kill().await;
        }
    }
}

/// Run a single Claude CLI invocation and stream events to `tx`.
/// Returns `true` if the process completed successfully.
async fn run_claude_process(
    sessions: &Arc<Mutex<HashMap<String, InternalSession>>>,
    session_id: &str,
    message: &str,
    path: &str,
    mode: PermissionMode,
    sdk_session_id: Option<&str>,
    tx: &mpsc::Sender<StreamEvent>,
) -> bool {
    // Build CLI arguments
    let mut args: Vec<String> = vec![
        "--print".to_string(),
        message.to_string(),
        "--output-format".to_string(),
        "stream-json".to_string(),
        "--verbose".to_string(),
    ];

    for flag in mode.to_cli_flags() {
        args.push(flag.to_string());
    }

    if let Some(sid) = sdk_session_id {
        args.push("--resume".to_string());
        args.push(sid.to_string());
    }

    info!("[SessionPool] Spawning claude for session {}", session_id);
    info!(
        "[SessionPool] Command: claude {}",
        args.iter()
            .map(|a| if a.contains(' ') {
                format!("\"{}\"", a)
            } else {
                a.clone()
            })
            .collect::<Vec<_>>()
            .join(" ")
    );
    info!("[SessionPool] CWD: {}", path);

    // Spawn Claude CLI process
    let child_result = Command::new("claude")
        .args(&args)
        .current_dir(path)
        .env("TERM", "dumb")
        .stdin(std::process::Stdio::null())
        .stdout(std::process::Stdio::piped())
        .stderr(std::process::Stdio::piped())
        .spawn();

    let mut child = match child_result {
        Ok(child) => child,
        Err(e) => {
            error!("[Session {}] Failed to spawn claude: {}", session_id, e);
            let _ = tx
                .send(StreamEvent::Error {
                    message: format!("Failed to spawn claude: {}", e),
                })
                .await;
            return false;
        }
    };

    // Take stdout and stderr before storing child
    let stdout = child.stdout.take();
    let stderr = child.stderr.take();

    // Store the child process in the session
    {
        let mut sessions_guard = sessions.lock().await;
        if let Some(session) = sessions_guard.get_mut(session_id) {
            session.process = Some(child);
        }
    }

    // Read stderr in a separate task (just logging)
    if let Some(stderr) = stderr {
        let sid = session_id.to_string();
        tokio::spawn(async move {
            let reader = BufReader::new(stderr);
            let mut lines = reader.lines();
            while let Ok(Some(line)) = lines.next_line().await {
                error!("[Session {}] stderr: {}", sid, line);
            }
        });
    }

    // Read stdout line-by-line (JSON-lines from Claude CLI)
    if let Some(stdout) = stdout {
        let reader = BufReader::new(stdout);
        let mut lines = reader.lines();

        while let Ok(Some(line)) = lines.next_line().await {
            let parsed: Value = match serde_json::from_str(&line) {
                Ok(v) => v,
                Err(_) => {
                    info!("[Session {}] non-JSON stdout: {}", session_id, line);
                    continue;
                }
            };

            // Extract model name from system init message
            if parsed.get("type").and_then(|v| v.as_str()) == Some("system")
                && parsed.get("subtype").and_then(|v| v.as_str()) == Some("init")
            {
                if let Some(model) = parsed.get("model").and_then(|v| v.as_str()) {
                    let mut sessions_guard = sessions.lock().await;
                    if let Some(session) = sessions_guard.get_mut(session_id) {
                        session.model = Some(model.to_string());
                    }
                    info!("[Session {}] Model: {}", session_id, model);
                }
            }

            // Convert to StreamEvent
            let event = convert_claude_message(&parsed);

            // Capture SDK session ID
            if let Some(sid) = event.session_id() {
                let mut sessions_guard = sessions.lock().await;
                if let Some(session) = sessions_guard.get_mut(session_id) {
                    session.sdk_session_id = Some(sid.to_string());
                }
            }

            // Send event to the channel
            if tx.send(event).await.is_err() {
                break;
            }
        }
    }

    // Wait for the process to exit
    let exit_status = {
        let mut sessions_guard = sessions.lock().await;
        if let Some(session) = sessions_guard.get_mut(session_id) {
            if let Some(ref mut child) = session.process {
                child.wait().await.ok()
            } else {
                None
            }
        } else {
            None
        }
    };

    // Check exit code — match TypeScript format: (code=N, signal=SIGNAME)
    if let Some(status) = exit_status {
        let code = status.code();

        #[cfg(unix)]
        let signal_str = {
            use std::os::unix::process::ExitStatusExt;
            status.signal().map(|s| format!("{}", s))
        };
        #[cfg(not(unix))]
        let signal_str: Option<String> = None;

        info!(
            "[Session {}] Process exited: code={}, signal={}",
            session_id,
            code.map_or("null".to_string(), |c| c.to_string()),
            signal_str.as_deref().unwrap_or("null"),
        );

        if code != Some(0) && code.is_some() {
            let _ = tx
                .send(StreamEvent::Error {
                    message: format!(
                        "Claude process exited abnormally (code={}, signal={})",
                        code.map_or("null".to_string(), |c| c.to_string()),
                        signal_str.as_deref().unwrap_or("null"),
                    ),
                })
                .await;
        }
    }

    // Kill process if still alive (e.g. on error/interrupt) — SIGTERM then SIGKILL after 3s
    {
        let mut sessions_guard = sessions.lock().await;
        if let Some(session) = sessions_guard.get_mut(session_id) {
            if let Some(ref mut child) = session.process {
                // Check if process is still alive by trying wait with timeout 0
                match child.try_wait() {
                    Ok(None) => {
                        // Still alive — send SIGTERM, then SIGKILL after 3s
                        send_sigterm_then_sigkill(child, 3000).await;
                    }
                    _ => {} // Already exited or error
                }
            }
        }
    }

    true
}

/// Process a message and then drain the queue (loop-based, no recursion).
/// This runs as a background tokio task.
async fn process_message_loop(
    sessions: Arc<Mutex<HashMap<String, InternalSession>>>,
    session_id: String,
    initial_message: String,
    initial_path: String,
    initial_mode: PermissionMode,
    initial_sdk_session_id: Option<String>,
    tx: mpsc::Sender<StreamEvent>,
) {
    // Process the initial message
    let success = run_claude_process(
        &sessions,
        &session_id,
        &initial_message,
        &initial_path,
        initial_mode,
        initial_sdk_session_id.as_deref(),
        &tx,
    )
    .await;

    if !success {
        let mut sessions_guard = sessions.lock().await;
        if let Some(session) = sessions_guard.get_mut(&session_id) {
            session.processing = false;
            session.status = SessionStatus::Idle;
            session.process = None;
        }
        return;
    }

    // Loop: reset state, check for queued messages, process them
    loop {
        let next_message = {
            let mut sessions_guard = sessions.lock().await;
            if let Some(session) = sessions_guard.get_mut(&session_id) {
                session.process = None;
                session.processing = false;
                session.status = SessionStatus::Idle;

                if session.queue.has_user_pending() {
                    session.queue.dequeue_user()
                } else {
                    None
                }
            } else {
                None
            }
        };

        let queued = match next_message {
            Some(q) => q,
            None => break, // No more queued messages
        };

        // Get current session state for the next message
        let (path, mode, sdk_sid) = {
            let sessions_guard = sessions.lock().await;
            if let Some(session) = sessions_guard.get(&session_id) {
                (
                    session.path.clone(),
                    session.mode,
                    session.sdk_session_id.clone(),
                )
            } else {
                break;
            }
        };

        // Mark as processing again
        {
            let mut sessions_guard = sessions.lock().await;
            if let Some(session) = sessions_guard.get_mut(&session_id) {
                session.processing = true;
                session.status = SessionStatus::Busy;
                session.last_activity_at = Utc::now();
            }
        }

        // For background queued messages, create a buffering channel
        let sessions_for_buffer = sessions.clone();
        let sid_for_buffer = session_id.clone();
        let (buf_tx, mut buf_rx) = mpsc::channel::<StreamEvent>(256);

        let buffer_task = tokio::spawn(async move {
            while let Some(event) = buf_rx.recv().await {
                let mut sessions_guard = sessions_for_buffer.lock().await;
                if let Some(session) = sessions_guard.get_mut(&sid_for_buffer) {
                    if !session.queue.is_client_connected() {
                        session.queue.buffer_response(event, false);
                    }
                }
            }
        });

        run_claude_process(
            &sessions,
            &session_id,
            &queued.message,
            &path,
            mode,
            sdk_sid.as_deref(),
            &buf_tx,
        )
        .await;

        drop(buf_tx);
        let _ = buffer_task.await;
    }
}

/// Expand ~ to home directory
fn expand_tilde(path: &str) -> String {
    if path.starts_with("~/") || path == "~" {
        if let Some(home) = dirs::home_dir() {
            return path.replacen('~', &home.to_string_lossy(), 1);
        }
    }
    path.to_string()
}
